Sources/GRPCHTTP2Core/Server/Connection/ServerConnectionManagementHandler.swift

```diff
@@ -0,0 +1,473 @@
+    maxIdleTime: TimeAmount?,
+    maxAge: TimeAmount?,
+    maxGraceTime: TimeAmount?,
+    keepAliveTime: TimeAmount?,
+    keepAliveTimeout: TimeAmount?,
+    allowKeepAliveWithoutCalls: Bool,
+    minPingIntervalWithoutCalls: TimeAmount,
+    clock: Clock = .nio
+  ) {
+    self.eventLoop = eventLoop
+
+    self.maxIdleTimer = maxIdleTime.map { Timer(delay: $0) }
+    self.maxAgeTimer = maxAge.map { Timer(delay: $0) }
+    self.maxGraceTimer = maxGraceTime.map { Timer(delay: $0) }
+
+    self.keepAliveTimer = keepAliveTime.map { Timer(delay: $0) }
+    // Always create a keep alive timeout timer, it's only use if there is a keep alive timer.
+    self.keepAliveTimeoutTimer = Timer(delay: keepAliveTimeout ?? .seconds(20))
+
+    // Generate a random value to be used as keep alive ping data.
+    let pingData = UInt64.random(in: .min ... .max)
+    self.keepAlivePingData = HTTP2PingData(withInteger: pingData)
+
+    self.state = StateMachine(
+      allowKeepAliveWithoutCalls: allowKeepAliveWithoutCalls,
+      minPingReceiveIntervalWithoutCalls: minPingIntervalWithoutCalls,
+      goAwayPingData: HTTP2PingData(withInteger: ~pingData)
+    )
+
+    self.flushPending = false
+    self.inReadLoop = false
+    self.clock = clock
+    self.frameStats = FrameStats()
+  }
+
+  func handlerAdded(context: ChannelHandlerContext) {
+    assert(context.eventLoop === self.eventLoop)
+  }
+
+  func channelActive(context: ChannelHandlerContext) {
+    self.maxAgeTimer?.schedule(on: context.eventLoop) {
+      self.initiateGracefulShutdown(context: context)
+    }
+
+    self.maxIdleTimer?.schedule(on: context.eventLoop) {
+      self.initiateGracefulShutdown(context: context)
+    }
+
+    self.keepAliveTimer?.schedule(on: context.eventLoop) {
+      self.keepAliveTimerFired(context: context)
+    }
+
+    context.fireChannelActive()
+  }
+
+  func channelInactive(context: ChannelHandlerContext) {
+    self.maxIdleTimer?.cancel()
+    self.maxAgeTimer?.cancel()
+    self.maxGraceTimer?.cancel()
+    self.keepAliveTimer?.cancel()
+    self.keepAliveTimeoutTimer.cancel()
+    context.fireChannelInactive()
+  }
+
+  func userInboundEventTriggered(context: ChannelHandlerContext, event: Any) {
+    switch event {
+    case let event as NIOHTTP2StreamCreatedEvent:
+      // The connection isn't idle if a stream is open.
+      self.maxIdleTimer?.cancel()
+      self.state.streamOpened(event.streamID)
+
+    case let event as StreamClosedEvent:
+      switch self.state.streamClosed(event.streamID) {
+      case .startIdleTimer:
+        self.maxIdleTimer?.schedule(on: context.eventLoop) {
+          self.initiateGracefulShutdown(context: context)
+        }
+
+      case .close:
+        context.close(mode: .all, promise: nil)
+
+      case .none:
+        ()
+      }
+
+    default:
+      ()
+    }
+
+    context.fireUserInboundEventTriggered(event)
+  }
+
+  func channelRead(context: ChannelHandlerContext, data: NIOAny) {
+    self.inReadLoop = true
```

Comment:
Adding this comment here but it applies to more of this PR. Why do we have some state management inside a state machine but that some state managed as vars of this handler. It isn't clear to me where that separation goes and it feels to me like we should have a single state machine for this handler that handles all the state.

Reply:
Layering, mostly. Most of the state within the handler that isn't the state machine are triggers which prod the state machine. Storing timers etc., back in the state machine complicates the state machine more than it helps. As for the flushing etc, it's not related to connection management, it's its own separate set of state.